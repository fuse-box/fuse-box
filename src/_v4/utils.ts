import * as fs from "fs";
import * as fsExtra from "fs-extra";
import { ensurePublicExtension, extractFuseBoxPath } from "../Utils";

const CACHED_PATHS = new Map<string, RegExp>();
export function path2Regex(path: string) {
	if (CACHED_PATHS.get(path)) {
		return CACHED_PATHS.get(path);
	}
	path = path.replace(/(\.|\/)/, "\\$1");

	const re = new RegExp(path);
	CACHED_PATHS.set(path, re);
	return re;
}

export function fileExists(file: string) {
	return fs.existsSync(file);
}

export function ensureDir(dir: string) {
	fsExtra.ensureDirSync(dir);
}
export function makeFuseBoxPath(homeDir: string, absPath: string) {
	return homeDir && ensurePublicExtension(extractFuseBoxPath(homeDir, absPath));
}
