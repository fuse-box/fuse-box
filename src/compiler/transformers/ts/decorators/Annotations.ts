import { ASTNode, ASTType } from '../../../interfaces/AST';

function id(name: string): ASTNode {
  return { type: 'Identifier', name };
}

export const KNOWN_IDENTIFIERS = {
  RegExp: 'RegExp',
  boolean: 'Boolean',
  Number: 'Number',
  String: 'String',
  Function: 'Function',
  Object: 'Object',

  Symbol: 'Symbol',
  Error: 'Error',
  EvalError: 'EvalError',
  RangeError: 'RangeError',
  ReferenceError: 'ReferenceError',
  SyntaxError: 'SyntaxError',
  TypeError: 'TypeError',
  URIError: 'URIError',
  Date: 'Date',
  Array: 'Array',
  Int8Array: 'Int8Array',
  Uint8Array: 'Uint8Array',
  Uint8ClampedArray: 'Uint8ClampedArray',
  Int16Array: 'Int16Array',
  Uint16Array: 'Uint16Array',
  Int32Array: 'Int32Array',
  Uint32Array: 'Uint32Array',
  Float32Array: 'Float32Array',
  Float64Array: 'Float64Array',
  BigInt64Array: 'BigInt64Array',
  BigUint64Array: 'BigUint64Array',
  Map: 'Map',
  Set: 'Set',
  WeakMap: 'WeakMap',
  WeakSet: 'WeakSet',
  ArrayBuffer: 'ArrayBuffer',
  DataView: 'DataView',
  Promise: 'Promise',
  VoidFunction: 'Function',
  GeneratorFunction: 'Function',
  FunctionConstructor: 'Function',
  FunctionStringCallback: 'Function',
  XMLHttpRequest: 'XMLHttpRequest',
  CallableFunction: 'Object',
  // here i give up with mapping. F.This
};

export const voidZero = {
  type: 'UnaryExpression',
  operator: 'void',
  argument: {
    type: 'Literal',
    value: 0,
  },
  prefix: true,
};

const id_Object = id('Object');
const id_String = id('String');
const id_Number = id('Number');
const id_Array = id('Array');
const id_Function = id('Function');
const id_Boolean = id('Boolean');

const LiteralTypes = {
  string: id_String,
  number: id_Number,
  boolean: id_Boolean,
  function: id_Function,
};

function maybe(id: string) {
  return {
    type: 'ConditionalExpression',
    test: {
      type: 'BinaryExpression',
      left: {
        type: 'UnaryExpression',
        operator: 'typeof',
        argument: {
          type: 'Identifier',
          name: id,
        },
        prefix: true,
      },
      right: {
        type: 'Literal',
        value: 'function',
      },
      operator: '===',
    },
    consequent: {
      type: 'Identifier',
      name: id,
    },
    alternate: id_Object,
  };
}

export function convertTypeAnnotation(node: ASTNode) {
  if (!node || node.type !== ASTType.TypeAnnotation) return id('Object');
  const typeAnnotation = node.typeAnnotation;

  switch (typeAnnotation.type) {
    case ASTType.BooleanKeyword:
      return id_Boolean;
    case ASTType.NeverKeyword:
    case ASTType.NullKeyword:
    case ASTType.VoidKeyword:
    case ASTType.UndefinedKeyword:
      return voidZero;
    case ASTType.AnyKeyword:
    case 'TypeOperator':
    case 'UnionType':
      return id_Object;
    case ASTType.StringKeyword:
      return id_String;
    case ASTType.NumberKeyword:
      return id_Number;
    case ASTType.FunctionType:
      return id_Function;
    case ASTType.TupleType:
    case ASTType.ArrayType:
      return id_Array;
    case ASTType.LiteralType:
      if (typeAnnotation.literal && typeAnnotation.literal.value !== undefined) {
        const t = typeof typeAnnotation.literal.value;

        if (LiteralTypes[t]) return LiteralTypes[t];
        else return id_Object;
      }
    case 'TypeLiteral':
      return id_Object;
    case ASTType.TypeReference:
      if (typeAnnotation.typeName && typeAnnotation.typeName.type === 'Identifier') {
        const known = KNOWN_IDENTIFIERS[typeAnnotation.typeName.name];
        if (known) return id(known);

        return maybe(typeAnnotation.typeName.name);
      }
      return id_Object;
    default:
      return id_Object;
  }
}
