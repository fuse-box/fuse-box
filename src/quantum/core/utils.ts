import * as path from "path";
import { joinFuseBoxPath } from "../../Utils";

export function generateFileCombinations(input: string): string[] {
    if (!input || input === ".") {
        return undefined;
    }
    let ext = path.extname(input);
    let combinations = [input];
    if (input.endsWith("/")) {
        combinations.push(
            joinFuseBoxPath(input, "index.js"),
            joinFuseBoxPath(input, "index.jsx")
        )
    } else {
        if (!ext) {
            combinations.push(
                `${input}.js`,
                `${input}.jsx`,
                joinFuseBoxPath(input, "index.js"),
                joinFuseBoxPath(input, "index.jsx"),
            )
        }
    }
    // some cases have weired conventions
    combinations.push(combinations + ".js");
    return combinations;
}
