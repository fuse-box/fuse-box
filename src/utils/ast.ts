export function nodeIsString(node) {
  return node.type === 'Literal' || node.type === 'StringLiteral';
}

interface IVisitProps {
  context?: IASTContext;
  idx?: number;
  parent?: any;
  prop?: string;
}
interface IASTContext {
  exports?: Array<string>;
  locals?: Array<string>;
}
interface IASTWalkProps {
  withScope?: boolean;
  visit?: (node: any, props: IVisitProps, context?: IASTContext) => void;
}
export function fastWalk(ast: any, walker: IASTWalkProps) {
  function visit(node, props: IVisitProps, context) {
    if (walker.withScope) {
      if (node.context) {
        context = node.context;
      }

      if (node.type === 'VariableDeclaration') {
        if (node.declarations) {
          for (const decl of node.declarations) {
            if (decl.type === 'VariableDeclarator' && decl.id && decl.id.type === 'Identifier') {
              if (context === undefined) context = { locals: [] };
              context.locals.push(decl.id.name);
              // we need to check for the next item on the list (if we are in an array)
              if (props.idx && props.prop) {
                if (props.parent[props.prop][props.idx + 1]) {
                  props.parent[props.prop][props.idx + 1].context = context;
                }
              }
            }
          }
        }
      }

      if (node.type === 'ExpressionStatement') {
        if (node.expression && node.expression.arguments) {
          node.expression.arguments.context = context;
        }
      }

      if (
        node.type === 'FunctionDeclaration' ||
        node.type === 'FunctionExpression' ||
        node.type === 'ArrowFunctionExpression'
      ) {
        if (node.id && node.id.type === 'Identifier') {
          // handles the following case:
          /*
            function o() {}
            console.log(o)
          */
          if (context === undefined) context = { locals: [] };
          context.locals.push(node.id.name);
          if (props.idx && props.prop) {
            if (props.parent[props.prop][props.idx + 1]) {
              props.parent[props.prop][props.idx + 1].context = context;
            }
          }
        }

        if (node.params) {
          for (const item of node.params) {
            if (item.type === 'Identifier') {
              let body = node.body;
              if (body.context === undefined) {
                // create context
                body.context = {
                  exports: context && context.exports ? context.exports.concat([]) : [],
                  // here we need to make a copy of the locals
                  locals: context ? context.locals.concat([]) : [],
                };
              }
              body.context.locals.push(item.name);
            }
          }
        }
      }
    }

    walker.visit(node, props, context);
    for (const prop in node) {
      if (prop[0] === '$') {
        continue;
      }
      const child = node[prop];
      if (child instanceof Array) {
        for (let i = 0; i < child.length; i++) {
          if (child && child[i] && child[i].type) {
            visit(child[i], { idx: i, parent: node, prop }, context);
          }
        }
      } else {
        if (child && child.type) visit(child, { parent: node, prop }, context);
      }
    }
  }

  visit(ast, {}, undefined);
}
