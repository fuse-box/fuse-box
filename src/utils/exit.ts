const handlers = {};
export function onExit(name: string, exitHandler: (type: string) => void) {
  handlers[name] = exitHandler;
}

function globalExitHandler(type: string, e) {
  for (const key in handlers) {
    handlers[key](type);
  }
  if (type === 'uncaughtException') {
    console.error(e);
  }

  if (type === 'exit') {
    process.exit();
  }
}
process.on('exit', e => globalExitHandler('cleanup', e));
process.on('SIGINT', e => globalExitHandler('exit', e));
process.on('SIGUSR1', e => globalExitHandler('exit', e));
process.on('SIGUSR2', e => globalExitHandler('exit', e));
process.on('uncaughtException', e => globalExitHandler('uncaughtException', e));
