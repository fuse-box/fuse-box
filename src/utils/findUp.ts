import { existsSync, statSync } from 'fs';
import { dirname, join } from 'path';

export const FIND_UP_GLOBAL_CONFIG = {
  maxStepsBack: 10,
};

export function findUp(
  start: string,
  target: string,
  boundaryArg?: {
    boundary: string;
    inclusive: boolean;
  },
): null | string {
  let currentDir = start;
  try {
    const stat = statSync(start);
    currentDir = stat.isDirectory() ? start : dirname(start);
  } catch (err) {}

  let lastTry = false;
  let backSteps = 0;
  while (backSteps++ <= FIND_UP_GLOBAL_CONFIG.maxStepsBack) {
    if (boundaryArg && boundaryArg.boundary.includes(currentDir)) {
      if (boundaryArg.inclusive && lastTry === false) {
        lastTry = true;
      } else {
        return null;
      }
    }

    const targetTestPath = join(currentDir, target);
    if (existsSync(targetTestPath)) {
      return targetTestPath;
    }

    currentDir = join(currentDir, '../');
  }

  console.error(`Too many back steps, starting from "${start}"`);
  return null;
}
