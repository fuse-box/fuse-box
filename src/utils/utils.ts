import * as appRoot from 'app-root-path';
import * as fs from 'fs';
import * as fsExtra from 'fs-extra';
import * as offsetLinesModule from 'offset-sourcemap-lines';
import * as path from 'path';
import { env } from '../env';
const CACHED_PATHS = new Map<string, RegExp>();
let prettyTime = require('pretty-time');

export function path2Regex(path: string) {
  if (CACHED_PATHS.get(path)) {
    return CACHED_PATHS.get(path);
  }
  path = path.replace(/(\.|\/)/, '\\$1');

  const re = new RegExp(path);
  CACHED_PATHS.set(path, re);
  return re;
}

export function matchAll(regex: RegExp, str: string, cb: (matches) => void) {
  let matches;
  while ((matches = regex.exec(str))) {
    cb(matches);
  }
}

export function getFileModificationTime(absPath) {
  return fs.statSync(absPath).mtime.getTime();
}

export function makePublicPath(target: string) {
  return ensureFuseBoxPath(path.relative(appRoot.path, target));
}

export function removeFolder(userPath) {
  fsExtra.removeSync(userPath);
}

export function readJSONFile(target: string) {
  return JSON.parse(readFile(target));
}

export function isPathRelative(from: string, to: string) {
  const relativePath = path.relative(from, to);
  return !relativePath.startsWith('..');
}

export function isDirectoryEmpty(directory: string) {
  const files = fs.readdirSync(directory);
  return files.length === 0;
}

export function getPublicPath(x: string) {
  return path.relative(env.APP_ROOT, x);
}
export function beautifyBundleName(absPath: string, maxLength?: number) {
  return absPath
    .replace(/(\.\w+)$/g, '')
    .split(/(\/|\\)/g)
    .filter(a => a !== '' && a !== '.' && !a.match(/(\/|\\)/g))
    .reduce((acc, curr, _idx, arr) =>
      acc ? (maxLength && acc.length > maxLength ? arr[arr.length - 1] : `${acc}-${curr}`) : curr,
    )
    .toLowerCase();
}

export function offsetLines(obj: any, amount: number) {
  return offsetLinesModule(obj, amount);
}

export function isRegExp(input: any): boolean {
  return !!(input && typeof input.test === 'function');
}

export function createRequireConst(name: string, variable?: string) {
  return `var ${variable ? variable : name} = require("${name}");`;
}
export function createRequireConstWithObject(name: string, variable: string, obj?: string) {
  return `var ${variable ? variable : name} = require("${name}").${obj};`;
}
export function createStringConst(name: string, value: string) {
  return `const ${name} = ${JSON.stringify(value)};`;
}

export function createVarString(name: string, value: string) {
  return `var ${name} = ${JSON.stringify(value)};`;
}
export function ensurePublicExtension(url: string) {
  let ext = path.extname(url);
  if (ext === '.ts') {
    url = replaceExt(url, '.js');
  }
  if (ext === '.tsx') {
    url = replaceExt(url, '.jsx');
  }
  return url;
}

export function parseVersion(version: string) {
  const re = /v?(\d+)/g;
  let matcher;
  const versions = [];
  while ((matcher = re.exec(version))) {
    versions.push(parseInt(matcher[1]));
  }
  return versions;
}

export function replaceExt(npath: string, ext): string {
  if (!npath) {
    return npath;
  }

  if (/\.[a-z0-9]+$/i.test(npath)) {
    return npath.replace(/\.[a-z0-9]+$/i, ext);
  } else {
    return npath + ext;
  }
}

export function extractFuseBoxPath(homeDir: string, targetPath: string) {
  homeDir = ensureFuseBoxPath(homeDir);
  targetPath = ensureFuseBoxPath(targetPath);
  let result = targetPath.replace(homeDir, '');
  if (result[0] === '/') {
    result = result.slice(1);
  }
  return result;
}

export const fileExists = fs.existsSync;

export function readFile(file: string) {
  return fs.readFileSync(file).toString();
}

export function readFileAsBuffer(file: string) {
  return fs.readFileSync(file);
}

export function removeFile(file: string) {
  return fs.unlinkSync(file);
}
export async function copyFile(file: string, target: string) {
  return fsExtra.copy(file, target);
}
export function isObject(obj: any) {
  return typeof obj === 'object';
}

export function pathJoin(...args) {
  return path.join(...args);
}
export function getExtension(file: string) {
  return path.extname(file);
}

export function ensureDir(dir: string) {
  fsExtra.ensureDirSync(dir);
  return dir;
}

export function fileStat(file: string) {
  return fs.statSync(file);
}
export function makeFuseBoxPath(homeDir: string, absPath: string) {
  return homeDir && ensurePublicExtension(extractFuseBoxPath(homeDir, absPath));
}

export function measureTime(group?: string) {
  let startTime = process.hrtime();
  return {
    end: (precision?) => {
      return prettyTime(process.hrtime(startTime), precision);
    },
  };
}

export function cleanExistingSourceMappingURL(contents: string) {
  return contents.replace(/\/*#\s*sourceMappingURL=\s*([^\s]+)\s*\*\//, '');
}

export function safeRegex(contents: string) {
  return new RegExp(contents.replace(/[-[\]{}()*+!<=:?.\/\\^$|#\s,]/g, '\\$&'), 'g');
}

export function findReplace(str: string, re: RegExp, fn: (args) => string) {
  return str.replace(re, (...args) => {
    return fn(args);
  });
}

export function path2RegexPattern(input: undefined | RegExp | string): RegExp {
  if (!input) {
    return;
  }

  if (typeof input === 'string') {
    let r = '';
    for (let i = 0; i < input.length; i++) {
      switch (input[i]) {
        case '*':
          r += '.*';
          break;
        case '.':
          r += '\\.';
          break;
        case '/':
          r += '(\\/|\\\\)';
          break;
        case '\\': // window paths
          r += '(\\/|\\\\)';
          break;
        default:
          r += input[i];
      }
    }
    return new RegExp(r);
  }
  return input;
}

export function ensureUserPath(userPath: string, root?: string) {
  if (!path.isAbsolute(userPath)) {
    userPath = path.join(root, userPath);
  }
  userPath = path.normalize(userPath);
  let dir = path.dirname(userPath);

  fsExtra.ensureDirSync(dir);
  return userPath;
}

export type Concat = {
  content: Buffer;
  sourceMap: string;
  add(fileName: null | string, content: Buffer | string, sourceMap?: string): void;
};
export type ConcatModule = {
  new (generateSourceMap: boolean, outputFileName: string, seperator: string): Concat;
};
export const Concat: ConcatModule = require('fuse-concat-with-sourcemaps');

export function createConcat(generateSourceMap: boolean, outputFileName: string, seperator: string): Concat {
  return new Concat(generateSourceMap, outputFileName, seperator);
}

export function ensureAbsolutePath(userPath: string, root: string) {
  if (!path.isAbsolute(userPath)) {
    return path.join(root, userPath);
  }
  return userPath;
}

export function ensureScriptRoot(userPath: string) {
  if (!path.isAbsolute(userPath)) {
    return path.join(env.SCRIPT_PATH, userPath);
  }
  return userPath;
}

export function getPathRelativeToConfig(props: { dirName: string; ensureDirExist?: boolean; fileName?: string }) {
  let target = props.fileName ? path.dirname(props.fileName) : props.dirName;
  const fileName = props.fileName && path.basename(props.fileName);
  if (!path.isAbsolute(target)) {
    target = path.join(env.SCRIPT_PATH, target);
  }
  if (props.ensureDirExist) {
    const baseDir = path.dirname(target);
    ensureDir(baseDir);
  }
  return fileName ? path.join(target, fileName) : target;
}

export function isNodeModuleInstalled(name) {
  try {
    return require(name);
  } catch (e) {
    return false;
  }
}

export function ensureFuseBoxPath(input: string) {
  return input.replace(/\\/g, '/').replace(/\/$/, '');
}

export function joinFuseBoxPath(...any): string {
  let includesProtocol = any[0].includes('://');
  let joinedPath = !includesProtocol
    ? path.join(...any)
    : any[0].replace(/([^/])$/, '$1/') + path.join(...any.slice(1));
  return ensureFuseBoxPath(joinedPath);
}

export async function writeFile(name: string, contents) {
  return new Promise((resolve, reject) => {
    ensureDir(path.dirname(name));
    fs.writeFile(name, contents, err => {
      if (err) return reject(err);
      return resolve();
    });
  });
}

export function fastHash(text: string): string {
  let hash = 0;
  if (text.length == 0) return '';
  for (let i = 0; i < text.length; i++) {
    let char = text.charCodeAt(i);
    hash = (hash << 5) - hash + char;
    hash = hash & hash; // Convert to 32bit integer
  }
  let result = hash.toString(16).toString();
  if (result.charAt(0) === '-') {
    result = result.replace(/-/, '0');
  }
  return result;
}
