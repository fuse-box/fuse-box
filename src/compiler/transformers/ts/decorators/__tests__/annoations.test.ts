import * as buntis from 'buntis';
import { convertTypeAnnotation, KNOWN_IDENTIFIERS } from '../Annotations';
import '../../../../../utils/test_utils';
function conv(typeAnnotation?: string) {
  const ast = buntis.parseTSModule(`function a(${typeAnnotation ? 'c : ' + typeAnnotation : 'c'}){}`, {
    directives: true,
    next: true,
  }) as any;
  //console.log(JSON.stringify(ast, null, 2));
  return convertTypeAnnotation(ast.body[0].params[0].typeAnnotation);
}

const generics = ['string', 'number', 'Number', 'String'];
const voidZero = ['undefined', 'never', 'void', 'null'];
const other = ['any', '{}', '[]', "'a'", '1', '() => {}', 'true', 'false', '0x22', 'keyof S'];

const KNOWN = Object.keys(KNOWN_IDENTIFIERS);

describe('Annotation test', () => {
  // it('should give an object without annotation', () => {
  //   const res = conv();
  //   expect(res).toMatchSnapshot();
  // });

  describe('other', () => {
    for (const name of other) {
      it(`should detect "${name}"`, () => {
        const res = conv(name);

        expect(JSON.stringify(res)).toMatchSnapshot();
      });
    }
  });

  describe('Void Zero', () => {
    for (const name of voidZero) {
      it(`should detect "${name}"`, () => {
        const res = conv(name);
        //console.log(name, '=>', res);
        expect(JSON.stringify(res)).toMatchSnapshot();
      });
    }
  });

  describe('Known identifiers', () => {
    for (const name of KNOWN) {
      it(`should detect "${name}"`, () => {
        const res = conv(name);
        expect(JSON.stringify(res)).toMatchSnapshot();
      });
    }
  });

  describe('Generics', () => {
    for (const name of generics) {
      it(`should detect "${name}"`, () => {
        const res = conv(name);
        //console.log(name, '=>', res);
        expect(JSON.stringify(res)).toMatchSnapshot();
      });
    }
  });

  describe('Maybe objects', () => {
    it(`should detect SomeObject as maybe 1`, () => {
      const res = conv('SomeStuff');

      expect(JSON.stringify(res)).toMatchSnapshot();
    });
  });
});
