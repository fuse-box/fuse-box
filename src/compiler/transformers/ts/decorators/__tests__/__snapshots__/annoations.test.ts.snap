// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Annotation test Generics should detect "Number" 1`] = `
Object {
  "name": "Number",
  "type": "Identifier",
}
`;

exports[`Annotation test Generics should detect "String" 1`] = `
Object {
  "name": "String",
  "type": "Identifier",
}
`;

exports[`Annotation test Generics should detect "number" 1`] = `
Object {
  "name": "Number",
  "type": "Identifier",
}
`;

exports[`Annotation test Generics should detect "string" 1`] = `
Object {
  "name": "String",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "Array" 1`] = `
Object {
  "name": "Array",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "ArrayBuffer" 1`] = `
Object {
  "name": "ArrayBuffer",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "BigInt64Array" 1`] = `
Object {
  "name": "BigInt64Array",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "BigUint64Array" 1`] = `
Object {
  "name": "BigUint64Array",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "CallableFunction" 1`] = `
Object {
  "name": "Object",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "DataView" 1`] = `
Object {
  "name": "DataView",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "Date" 1`] = `
Object {
  "name": "Date",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "Error" 1`] = `
Object {
  "name": "Error",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "EvalError" 1`] = `
Object {
  "name": "EvalError",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "Float32Array" 1`] = `
Object {
  "name": "Float32Array",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "Float64Array" 1`] = `
Object {
  "name": "Float64Array",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "Function" 1`] = `
Object {
  "name": "Function",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "FunctionConstructor" 1`] = `
Object {
  "name": "Function",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "FunctionStringCallback" 1`] = `
Object {
  "name": "Function",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "GeneratorFunction" 1`] = `
Object {
  "name": "Function",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "Int8Array" 1`] = `
Object {
  "name": "Int8Array",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "Int16Array" 1`] = `
Object {
  "name": "Int16Array",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "Int32Array" 1`] = `
Object {
  "name": "Int32Array",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "Map" 1`] = `
Object {
  "name": "Map",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "Number" 1`] = `
Object {
  "name": "Number",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "Object" 1`] = `
Object {
  "name": "Object",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "Promise" 1`] = `
Object {
  "name": "Promise",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "RangeError" 1`] = `
Object {
  "name": "RangeError",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "ReferenceError" 1`] = `
Object {
  "name": "ReferenceError",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "RegExp" 1`] = `
Object {
  "name": "RegExp",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "Set" 1`] = `
Object {
  "name": "Set",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "String" 1`] = `
Object {
  "name": "String",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "Symbol" 1`] = `
Object {
  "name": "Symbol",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "SyntaxError" 1`] = `
Object {
  "name": "SyntaxError",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "TypeError" 1`] = `
Object {
  "name": "TypeError",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "URIError" 1`] = `
Object {
  "name": "URIError",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "Uint8Array" 1`] = `
Object {
  "name": "Uint8Array",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "Uint8ClampedArray" 1`] = `
Object {
  "name": "Uint8ClampedArray",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "Uint16Array" 1`] = `
Object {
  "name": "Uint16Array",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "Uint32Array" 1`] = `
Object {
  "name": "Uint32Array",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "VoidFunction" 1`] = `
Object {
  "name": "Function",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "WeakMap" 1`] = `
Object {
  "name": "WeakMap",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "WeakSet" 1`] = `
Object {
  "name": "WeakSet",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "XMLHttpRequest" 1`] = `
Object {
  "name": "XMLHttpRequest",
  "type": "Identifier",
}
`;

exports[`Annotation test Known identifiers should detect "boolean" 1`] = `undefined`;

exports[`Annotation test Void Zero should detect "never" 1`] = `
Object {
  "argument": Object {
    "type": "Literal",
    "value": 0,
  },
  "operator": "void",
  "prefix": true,
  "type": "UnaryExpression",
}
`;

exports[`Annotation test Void Zero should detect "null" 1`] = `
Object {
  "argument": Object {
    "type": "Literal",
    "value": 0,
  },
  "operator": "void",
  "prefix": true,
  "type": "UnaryExpression",
}
`;

exports[`Annotation test Void Zero should detect "undefined" 1`] = `
Object {
  "argument": Object {
    "type": "Literal",
    "value": 0,
  },
  "operator": "void",
  "prefix": true,
  "type": "UnaryExpression",
}
`;

exports[`Annotation test Void Zero should detect "void" 1`] = `
Object {
  "argument": Object {
    "type": "Literal",
    "value": 0,
  },
  "operator": "void",
  "prefix": true,
  "type": "UnaryExpression",
}
`;

exports[`Annotation test other should detect "'a'" 1`] = `
Object {
  "name": "String",
  "type": "Identifier",
}
`;

exports[`Annotation test other should detect "() => {}" 1`] = `
Object {
  "name": "Function",
  "type": "Identifier",
}
`;

exports[`Annotation test other should detect "[]" 1`] = `
Object {
  "name": "Array",
  "type": "Identifier",
}
`;

exports[`Annotation test other should detect "{}" 1`] = `
Object {
  "name": "Object",
  "type": "Identifier",
}
`;

exports[`Annotation test other should detect "0x22" 1`] = `
Object {
  "name": "Number",
  "type": "Identifier",
}
`;

exports[`Annotation test other should detect "1" 1`] = `
Object {
  "name": "Number",
  "type": "Identifier",
}
`;

exports[`Annotation test other should detect "any" 1`] = `
Object {
  "name": "Object",
  "type": "Identifier",
}
`;

exports[`Annotation test other should detect "false" 1`] = `
Object {
  "name": "Boolean",
  "type": "Identifier",
}
`;

exports[`Annotation test other should detect "keyof S" 1`] = `
Object {
  "name": "Object",
  "type": "Identifier",
}
`;

exports[`Annotation test other should detect "true" 1`] = `
Object {
  "name": "Boolean",
  "type": "Identifier",
}
`;
